3장. 데이터의 첫인상 읽기

In [ ]:

0. 제출 정보
이름: 예민호
GitHub ID: yegamanse-cpu
작성일: 2026.09.22
최종 제출 URL: https://github.com/yegamanse-cpu/llm-data-analysis-study/blob/main/chapter03/chapter03.ipynb

1. 데이터 로딩과 구조 확인

In [33]:
from pathlib import Path

import pandas as pd

pd.set_option("display.max_columns", 50)
pd.set_option("display.width", 120)


def find_project_root(start: Path) -> Path:
    """현재 위치에서 위로 올라가며 data/raw 폴더가 있는 프로젝트 루트를 찾습니다."""
    start = start.resolve()
    for candidate in [start, *start.parents]:
        if (candidate / "data" / "raw").exists() and (candidate / "book").exists():
            return candidate
    raise FileNotFoundError("프로젝트 루트를 찾지 못했습니다. 노트북을 저장소 안에서 실행해 주세요.")


PROJECT_ROOT = find_project_root(Path.cwd())
DATA_DIR = PROJECT_ROOT / "data" / "raw"

print("프로젝트 루트:", PROJECT_ROOT)
print("데이터 폴더:", DATA_DIR)

import sys

print(sys.executable)
print(Path.cwd())

프로젝트 루트: C:\dev\llm-data-analysis-course
데이터 폴더: C:\dev\llm-data-analysis-course\data\raw
c:\dev\llm-data-analysis-course\.venv\Scripts\python.exe
c:\dev\llm-data-analysis-course\notebooks


In [2]:
expected_files = [
    "customers.csv",
    "products.csv",
    "orders.csv",
    "order_items.csv",
]

file_check = pd.DataFrame({
    "file": expected_files,
    "path": [str(DATA_DIR / filename) for filename in expected_files],
    "exists": [(DATA_DIR / filename).exists() for filename in expected_files],
})

file_check

,file,path,exists
0,customers.csv,C:\dev\llm-data-analysis-course\data\raw\custo...,True
1,products.csv,C:\dev\llm-data-analysis-course\data\raw\produ...,True
2,orders.csv,C:\dev\llm-data-analysis-course\data\raw\order...,True
3,order_items.csv,C:\dev\llm-data-analysis-course\data\raw\order...,True


In [34]:
expected_files = [
    "customers.csv",
    "products.csv",
    "orders.csv",
    "order_items.csv",
]

file_check = {
    filename: (DATA_DIR / filename).exists()
    for filename in expected_files
}

file_check

{'customers.csv': True,
 'products.csv': True,
 'orders.csv': True,
 'order_items.csv': True}

In [3]:
customers = pd.read_csv(DATA_DIR / "customers.csv")
products = pd.read_csv(DATA_DIR / "products.csv")
orders = pd.read_csv(DATA_DIR / "orders.csv")
order_items = pd.read_csv(DATA_DIR / "order_items.csv")

print("customers:", type(customers))
print("products:", type(products))
print("orders:", type(orders))
print("order_items:", type(order_items))

customers: <class 'pandas.DataFrame'>
products: <class 'pandas.DataFrame'>
orders: <class 'pandas.DataFrame'>
order_items: <class 'pandas.DataFrame'>


In [4]:
datasets = {
    "customers": customers,
    "products": products,
    "orders": orders,
    "order_items": order_items,
}

list(datasets.keys())

['customers', 'products', 'orders', 'order_items']

In [5]:
print("customers:", customers.shape)
print("products:", products.shape)
print("orders:", orders.shape)
print("order_items:", order_items.shape)

customers: (150, 6)
products: (100, 4)
orders: (300, 5)
order_items: (764, 5)


In [6]:
shape_summary = pd.DataFrame([
    {
        "dataset": name,
        "rows": df.shape[0],
        "columns": df.shape[1],
    }
    for name, df in datasets.items()
])

shape_summary

,dataset,rows,columns
0,customers,150,6
1,products,100,4
2,orders,300,5
3,order_items,764,5


In [7]:
customers.head()

,customer_id,name,gender,age,city,signup_date
0,1,김수민,F,19,광주,2024-08-26
1,2,김정호,F,32,대구,2026-01-09
2,3,이경수,F,61,성남,2024-08-19
3,4,조영호,F,55,울산,2026-06-20
4,5,이예원,F,19,부산,2024-11-20


In [8]:
products.head()

,product_id,product_name,category,price
0,1,전자기기 상품 001,전자기기,160000
1,2,도서 상품 002,도서,34000
2,3,전자기기 상품 003,전자기기,152000
3,4,생활용품 상품 004,생활용품,70000
4,5,식품 상품 005,식품,186000


In [9]:
orders.head()

,order_id,customer_id,order_date,payment_method,order_status
0,1,123,2026-07-14,card,completed
1,2,77,2025-09-29,naver_pay,cancelled
2,3,138,2026-01-26,bank_transfer,cancelled
3,4,57,2026-04-08,kakao_pay,cancelled
4,5,125,2026-02-27,card,cancelled


In [10]:
order_items.head()

,order_item_id,order_id,product_id,quantity,unit_price
0,1,1,100,3,102000
1,2,1,87,5,25000
2,3,1,7,3,142000
3,4,1,9,3,193000
4,5,2,72,4,189000


In [11]:
customers.tail()

,customer_id,name,gender,age,city,signup_date
145,146,김숙자,M,61,성남,2026-03-01
146,147,이정남,M,19,부산,2025-04-20
147,148,오도현,M,29,고양,2026-08-22
148,149,김정자,M,20,부산,2024-12-26
149,150,조미영,M,40,대전,2026-02-10


In [35]:
customers.head()
customers.tail()
products.head()
orders.head()
order_items.head()

,order_item_id,order_id,product_id,quantity,unit_price
0,1,1,100,3,102000
1,2,1,87,5,25000
2,3,1,7,3,142000
3,4,1,9,3,193000
4,5,2,72,4,189000


In [12]:
for name, df in datasets.items():
    print(f"[{name}]")
    print(list(df.columns))
    print()

[customers]
['customer_id', 'name', 'gender', 'age', 'city', 'signup_date']

[products]
['product_id', 'product_name', 'category', 'price']

[orders]
['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status']

[order_items]
['order_item_id', 'order_id', 'product_id', 'quantity', 'unit_price']



In [13]:
column_summary = pd.DataFrame([
    {
        "dataset": name,
        "column_count": len(df.columns),
        "column_names": ", ".join(df.columns),
    }
    for name, df in datasets.items()
])

column_summary

,dataset,column_count,column_names
0,customers,6,"customer_id, name, gender, age, city, signup_date"
1,products,4,"product_id, product_name, category, price"
2,orders,5,"order_id, customer_id, order_date, payment_met..."
3,order_items,5,"order_item_id, order_id, product_id, quantity,..."


In [14]:
customers.info()

<class 'pandas.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   customer_id  150 non-null    int64
 1   name         150 non-null    str  
 2   gender       150 non-null    str  
 3   age          150 non-null    int64
 4   city         150 non-null    str  
 5   signup_date  150 non-null    str  
dtypes: int64(2), str(4)
memory usage: 7.2 KB


In [15]:
for name, df in datasets.items():
    print(f"\n===== {name} =====")
    df.info()


===== customers =====
<class 'pandas.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   customer_id  150 non-null    int64
 1   name         150 non-null    str  
 2   gender       150 non-null    str  
 3   age          150 non-null    int64
 4   city         150 non-null    str  
 5   signup_date  150 non-null    str  
dtypes: int64(2), str(4)
memory usage: 7.2 KB

===== products =====
<class 'pandas.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   product_id    100 non-null    int64
 1   product_name  100 non-null    str  
 2   category      100 non-null    str  
 3   price         100 non-null    int64
dtypes: int64(2), str(2)
memory usage: 3.3 KB

===== orders =====
<class 'pandas.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 5 columns):

In [16]:
dtype_summary = pd.concat(
    [df.dtypes.rename(name) for name, df in datasets.items()],
    axis=1,
).fillna("")

dtype_summary

,customers,products,orders,order_items
customer_id,int64,,int64,
name,str,,,
gender,str,,,
age,int64,,,
city,str,,,
signup_date,str,,,
product_id,,int64,,int64
product_name,,str,,
category,,str,,
price,,int64,,


실행/결과
-4개 CSV 로딩 여부: [x]

-각 데이터 shape: 
customers: (150, 6)
products: (100, 4)
orders: (300, 5)
order_items: (764, 5)

-주요 컬럼: customer_id, product_id, order_id

-dtypes에서 주목한 컬럼: customer_id, product_id, order_id


Evidence
![](images/01.png)

결과 관찰 
- 각 파일을 로드하고 데이터의 shape의 수(100~764, 4~6)와 colums명(customer_id 등) dtype(int64, str)을 확인함

나의 해석과 판단 
- 주요 칼럼인 customer_id, product_id, order_id가 각 raw데이터의 행과 연결되어 있다. 데이터 분석의 오류를 막기 위해 부모-자식 데이터 간의 관계설정이 중요하다고 판단되는 부분이다.

업무·분석적 의미 
- raw데이터를 취합하면 서로 연결된 customer_id, product_id, order_id간의 데이터가 중복되어 오류가 생길 수 있을 것으로 판단된다. 또한 dtype이 int64인지 str인지도 미리 파악하고 있어야 함수사용 시 오류가 발생하지 않을 것으로 판단된다.

한계와 추가 확인 사항 아직 
- 다만, 아직 데이터가 취합된 형태는 아니기 때문에 예상되는 문제점이 실제 발생할 지는 확인이 필요하다.

2. 결측·중복·키 품질

In [17]:
customers.isna().sum()

customer_id    0
name           0
gender         0
age            0
city           0
signup_date    0
dtype: int64

In [36]:
for name, df in datasets.items():
    print(f"\n[{name}]")
    print(df.isna().sum())


[customers]
customer_id    0
name           0
gender         0
age            0
city           0
signup_date    0
dtype: int64

[products]
product_id      0
product_name    0
category        0
price           0
dtype: int64

[orders]
order_id          0
customer_id       0
order_date        0
payment_method    0
order_status      0
dtype: int64

[order_items]
order_item_id    0
order_id         0
product_id       0
quantity         0
unit_price       0
dtype: int64


In [37]:
(customers.isna().mean() * 100).round(2)


customer_id    0.0
name           0.0
gender         0.0
age            0.0
city           0.0
signup_date    0.0
dtype: float64

In [18]:
missing_summary = pd.concat(
    [df.isna().sum().rename(name) for name, df in datasets.items()],
    axis=1,
).fillna("").astype(str)

missing_summary

,customers,products,orders,order_items
customer_id,0.0,,0.0,
name,0.0,,,
gender,0.0,,,
age,0.0,,,
city,0.0,,,
signup_date,0.0,,,
product_id,,0.0,,0.0
product_name,,0.0,,
category,,0.0,,
price,,0.0,,


In [19]:
missing_rate_summary = pd.concat(
    [(df.isna().mean() * 100).round(2).rename(name) for name, df in datasets.items()],
    axis=1,
).fillna("")

missing_rate_summary

,customers,products,orders,order_items
customer_id,0.0,,0.0,
name,0.0,,,
gender,0.0,,,
age,0.0,,,
city,0.0,,,
signup_date,0.0,,,
product_id,,0.0,,0.0
product_name,,0.0,,
category,,0.0,,
price,,0.0,,


In [20]:
duplicate_rows = pd.DataFrame([
    {
        "dataset": name,
        "duplicated_rows": df.duplicated().sum(),
    }
    for name, df in datasets.items()
])

duplicate_rows

,dataset,duplicated_rows
0,customers,0
1,products,0
2,orders,0
3,order_items,0


In [21]:
id_duplicate_checks = pd.DataFrame([
    {
        "check": "customers.customer_id",
        "duplicated_count": customers["customer_id"].duplicated().sum(),
        "interpretation": "0이어야 고객 ID가 유일합니다.",
    },
    {
        "check": "products.product_id",
        "duplicated_count": products["product_id"].duplicated().sum(),
        "interpretation": "0이어야 상품 ID가 유일합니다.",
    },
    {
        "check": "orders.order_id",
        "duplicated_count": orders["order_id"].duplicated().sum(),
        "interpretation": "0이어야 주문 ID가 유일합니다.",
    },
    {
        "check": "order_items.order_id",
        "duplicated_count": order_items["order_id"].duplicated().sum(),
        "interpretation": "한 주문에 여러 상품이 있으면 0보다 클 수 있습니다.",
    },
])

id_duplicate_checks

,check,duplicated_count,interpretation
0,customers.customer_id,0,0이어야 고객 ID가 유일합니다.
1,products.product_id,0,0이어야 상품 ID가 유일합니다.
2,orders.order_id,0,0이어야 주문 ID가 유일합니다.
3,order_items.order_id,464,한 주문에 여러 상품이 있으면 0보다 클 수 있습니다.


In [38]:
for name, df in datasets.items():
    print(name, df.duplicated().sum())

customers 0
products 0
orders 0
order_items 0


In [39]:
key_columns = {
    "customers": "customer_id",
    "products": "product_id",
    "orders": "order_id",
    "order_items": "order_item_id",
}

for name, key in key_columns.items():
    df = datasets[name]
    print(
        name,
        "missing:", df[key].isna().sum(),
        "duplicated:", df[key].duplicated().sum(),
    )

customers missing: 0 duplicated: 0
products missing: 0 duplicated: 0
orders missing: 0 duplicated: 0
order_items missing: 0 duplicated: 0


-주요 ID 결측: [X]
-주요 ID 중복: [X]
-전체 행 중복: [X]

Evidence
![](images/02.png)

결과 관찰 
- isna 및 duplicate로 확인한 결과, 결측 및 중복이 없는 것으로 확인되었다. 결측치는 앞서 info로 기 확인하였다.

나의 해석과 판단 
- 데이터를 확인하고 결측 및 중복데이터가 있을 경우 분석오류(runtime-error)가 발생할 수 있으므로,
- 데이터를 "0"으로 넣을지 보정치를 넣을지(결측치) 또는 삭제(중복데이터)를 할 지의 의사결정을 한 후 데이터 분석을 해야한다고 생각한다.

업무·분석적 의미 
- 분석오류(runtime-error)가 발생할 때마다 일일이 수정하기 보다 분석초기에 구조를 먼저 파악하고 미리 보정해야 원활하게 분석이 추진될 수 있다.

한계와 추가 확인 사항 아직 
- customers.customer_id와 order_items.order_id의 숫자가 다른 이유는 customers - customer_id는 서로 일치된 고유 값을 가지기 때문에 중복이 발생 할 수 없으나,
- order_items - order_id는 order_id에 여러 order_items가 있을 수 있으니 중복이 발생하는 것이 자연스럽다.

3. 숫자형·범주형·날짜 점검

In [22]:
customers.describe()

,customer_id,age
count,150.000000,150.000000
mean,75.500000,42.086667
std,43.445368,15.613166
min,1.000000,19.000000
25%,38.250000,29.000000
50%,75.500000,40.000000
75%,112.750000,57.000000
max,150.000000,69.000000


In [23]:
products[["price"]].describe()

,price
count,100.000000
mean,110040.000000
std,56433.910574
min,5000.000000
25%,65750.000000
50%,112000.000000
75%,161000.000000
max,200000.000000


In [24]:
order_items[["quantity", "unit_price"]].describe()

,quantity,unit_price
count,764.000000,764.000000
mean,3.053665,108561.518325
std,1.410873,56996.770604
min,1.000000,5000.000000
25%,2.000000,62000.000000
50%,3.000000,111000.000000
75%,4.000000,161250.000000
max,5.000000,200000.000000


In [40]:
customers[["age"]].describe()
products[["price"]].describe()
order_items[["quantity", "unit_price"]].describe()

,quantity,unit_price
count,764.000000,764.000000
mean,3.053665,108561.518325
std,1.410873,56996.770604
min,1.000000,5000.000000
25%,2.000000,62000.000000
50%,3.000000,111000.000000
75%,4.000000,161250.000000
max,5.000000,200000.000000


In [25]:
price_text = pd.Series(["10,000", "25,500", "3000", "확인필요"])
price_number = pd.to_numeric(
    price_text.str.replace(",", "", regex=False),
    errors="coerce",
)

pd.DataFrame({
    "original": price_text,
    "converted": price_number,
})

,original,converted
0,"10,000",10000.0
1,"25,500",25500.0
2,3000,3000.0
3,확인필요,NaN


In [26]:
customers["city"].value_counts().head(10)

city
성남    21
광주    17
부산    16
대구    15
서울    15
울산    14
인천    14
대전    14
수원    13
고양    11
Name: count, dtype: int64

In [27]:
products["category"].value_counts()

category
스포츠     19
전자기기    17
생활용품    16
뷰티      16
도서      14
패션      11
식품       7
Name: count, dtype: int64

In [28]:
orders["order_status"].value_counts()

order_status
completed    184
cancelled     64
refunded      52
Name: count, dtype: int64

In [41]:
customers["city"].value_counts(dropna=False)
products["category"].value_counts(dropna=False)
orders["order_status"].value_counts(dropna=False)

order_status
completed    184
cancelled     64
refunded      52
Name: count, dtype: int64

In [29]:
categorical_summary = pd.DataFrame([
    {"dataset": "customers", "column": "city", "unique_count": customers["city"].nunique()},
    {"dataset": "customers", "column": "gender", "unique_count": customers["gender"].nunique()},
    {"dataset": "products", "column": "category", "unique_count": products["category"].nunique()},
    {"dataset": "orders", "column": "payment_method", "unique_count": orders["payment_method"].nunique()},
    {"dataset": "orders", "column": "order_status", "unique_count": orders["order_status"].nunique()},
])

categorical_summary

,dataset,column,unique_count
0,customers,city,10
1,customers,gender,2
2,products,category,7
3,orders,payment_method,4
4,orders,order_status,3


In [30]:
orders["order_date"].head()

0    2026-07-14
1    2025-09-29
2    2026-01-26
3    2026-04-08
4    2026-02-27
Name: order_date, dtype: str

In [42]:
orders["order_date"] = pd.to_datetime(
    orders["order_date"],
    errors="coerce",
)

print("변환 실패:", orders["order_date"].isna().sum())
print("가장 빠른 주문일:", orders["order_date"].min())
print("가장 최근 주문일:", orders["order_date"].max())

변환 실패: 0
가장 빠른 주문일: 2025-09-15 00:00:00
가장 최근 주문일: 2026-09-14 00:00:00


In [43]:
signup_check = pd.to_datetime(
    customers["signup_date"],
    errors="coerce",
)

print("가입일 변환 실패:", signup_check.isna().sum())

가입일 변환 실패: 0


In [31]:
print("변환 전 타입:", orders["order_date"].dtype)

orders["order_date"] = pd.to_datetime(orders["order_date"], errors="coerce")

print("변환 후 타입:", orders["order_date"].dtype)
print("날짜 변환 실패 건수:", orders["order_date"].isna().sum())
print("가장 빠른 주문일:", orders["order_date"].min())
print("가장 최근 주문일:", orders["order_date"].max())

변환 전 타입: str
변환 후 타입: datetime64[us]
날짜 변환 실패 건수: 0
가장 빠른 주문일: 2025-09-15 00:00:00
가장 최근 주문일: 2026-09-14 00:00:00


In [32]:
orders.assign(
    order_year=orders["order_date"].dt.year,
    order_month=orders["order_date"].dt.month,
    order_day_name=orders["order_date"].dt.day_name(),
).head()

,order_id,customer_id,order_date,payment_method,order_status,order_year,order_month,order_day_name
0,1,123,2026-07-14,card,completed,2026,7,Tuesday
1,2,77,2025-09-29,naver_pay,cancelled,2025,9,Monday
2,3,138,2026-01-26,bank_transfer,cancelled,2026,1,Monday
3,4,57,2026-04-08,kakao_pay,cancelled,2026,4,Wednesday
4,5,125,2026-02-27,card,cancelled,2026,2,Friday


-숫자형 범위에서 주목한 값: price
-범주형 빈도에서 주목한 값: city
-날짜 변환 실패 건수: 0
-날짜 범위: 2025-09-15 ~ 2026-09-14

Evidence
![](images/03.png)

결과 관찰 
- age, price, quantity, unit_price, city, category, order_status의 범위 및 빈도를 확인하였다. 날짜 변환건수는 없었으며 price에서 문자로 되어 있는 부분을 숫자로 변환하였다.

나의 해석과 판단 
- 앞선 실습과정에서 나와 있듯 문자로 되어 있는자료가 데이터를 육안으로 살펴보면 잘 보이지 않기 때문에 전처리 과정에서 조치를 하고 분석이 필요하다.

업무·분석적 의미 
- 실무에서 데이터를 추출하다보면 위와 같은 상황들이 많이 발생한다. 전처리를 했다고 하더라도 일부 결과들이 누락되는 일이 종종 발생하기 때문에 전치리 과정이 매우 중요하다.

한계와 추가 확인 사항 
- 첫번째 실습과정에서 언급한 내용으로 날짜범위가 2025.9~2026.09로 전년대비 매출의 변화나 원인 등을 파악할 충분한 수치는 아니라고 생각한다.


4. CSV 간 키 관계 검증

In [44]:
invalid_customers = orders[~orders["customer_id"].isin(customers["customer_id"])]
invalid_orders = order_items[~order_items["order_id"].isin(orders["order_id"])]
invalid_products = order_items[~order_items["product_id"].isin(products["product_id"])]

relationship_check = pd.DataFrame([
    {
        "relationship": "orders.customer_id -> customers.customer_id",
        "invalid_rows": len(invalid_customers),
    },
    {
        "relationship": "order_items.order_id -> orders.order_id",
        "invalid_rows": len(invalid_orders),
    },
    {
        "relationship": "order_items.product_id -> products.product_id",
        "invalid_rows": len(invalid_products),
    },
])

relationship_check

,relationship,invalid_rows
0,orders.customer_id -> customers.customer_id,0
1,order_items.order_id -> orders.order_id,0
2,order_items.product_id -> products.product_id,0


In [45]:
invalid_customers = orders.loc[
    ~orders["customer_id"].isin(customers["customer_id"])
]

invalid_orders = order_items.loc[
    ~order_items["order_id"].isin(orders["order_id"])
]

invalid_products = order_items.loc[
    ~order_items["product_id"].isin(products["product_id"])
]

print("없는 customer_id:", len(invalid_customers))
print("없는 order_id:", len(invalid_orders))
print("없는 product_id:", len(invalid_products))

없는 customer_id: 0
없는 order_id: 0
없는 product_id: 0


In [46]:
order_items_with_products = order_items.merge(
    products,
    on="product_id",
    how="left",
)

order_items_with_products["line_amount"] = (
    order_items_with_products["quantity"] * order_items_with_products["unit_price"]
)

order_items_with_products.head()

,order_item_id,order_id,product_id,quantity,unit_price,product_name,category,price,line_amount
0,1,1,100,3,102000,도서 상품 100,도서,102000,306000
1,2,1,87,5,25000,도서 상품 087,도서,25000,125000
2,3,1,7,3,142000,도서 상품 007,도서,142000,426000
3,4,1,9,3,193000,스포츠 상품 009,스포츠,193000,579000
4,5,2,72,4,189000,뷰티 상품 072,뷰티,189000,756000


In [47]:
category_sales = (
    order_items_with_products
    .groupby("category", as_index=False)["line_amount"]
    .sum()
    .sort_values("line_amount", ascending=False)
)

category_sales

,category,line_amount
3,스포츠,50174000
1,뷰티,47551000
5,전자기기,41003000
2,생활용품,34839000
4,식품,33597000
0,도서,24645000
6,패션,23801000


In [48]:
def check_data_overview(name: str, df: pd.DataFrame) -> None:
    print(f"===== {name} =====")
    print("shape:", df.shape)
    print("\ncolumns:")
    print(list(df.columns))
    print("\ndtypes:")
    print(df.dtypes)
    print("\nmissing values:")
    print(df.isna().sum())
    print("\nduplicated rows:", df.duplicated().sum())


check_data_overview("customers", customers)

===== customers =====
shape: (150, 6)

columns:
['customer_id', 'name', 'gender', 'age', 'city', 'signup_date']

dtypes:
customer_id    int64
name             str
gender           str
age            int64
city             str
signup_date      str
dtype: object

missing values:
customer_id    0
name           0
gender         0
age            0
city           0
signup_date    0
dtype: int64

duplicated rows: 0


In [49]:
for name, df in datasets.items():
    check_data_overview(name, df)
    print()

===== customers =====
shape: (150, 6)

columns:
['customer_id', 'name', 'gender', 'age', 'city', 'signup_date']

dtypes:
customer_id    int64
name             str
gender           str
age            int64
city             str
signup_date      str
dtype: object

missing values:
customer_id    0
name           0
gender         0
age            0
city           0
signup_date    0
dtype: int64

duplicated rows: 0

===== products =====
shape: (100, 4)

columns:
['product_id', 'product_name', 'category', 'price']

dtypes:
product_id      int64
product_name      str
category          str
price           int64
dtype: object

missing values:
product_id      0
product_name    0
category        0
price           0
dtype: int64

duplicated rows: 0

===== orders =====
shape: (300, 5)

columns:
['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status']

dtypes:
order_id                   int64
customer_id                int64
order_date        datetime64[us]
payment_method          

-없는 customer_id: 0

-없는 order_id: 0

-없는 product_id: 0


Evidence
![](images/04.png)

결과 관찰 
- orders.customer_id -> customers.customer_id /	order_items.order_id -> orders.order_id	/ order_items.product_id -> products.product_id	이 모두 0이며 invalid_rows이 모두 0으로 기본적인 연결 관계가 유지되고 있다고 판단된다.

나의 해석과 판단 
- 데이터 간 연결을 통하여 각 파일들이 가지고 있던 데이터를 종합하여 분석할 수 있게 되었다. 

업무·분석적 의미 
- 데이터는 각자 가지고는 칼럼의 정보를 가지고 있지만 이들이 모두 연결되어야 보다 의미있는 분석이 가능함을 확인하였다.


5. LLM 구조 설명 검증

In [50]:
llm_dataset_summary = shape_summary.merge(column_summary, on="dataset")
llm_dataset_summary

,dataset,rows,columns,column_count,column_names
0,customers,150,6,6,"customer_id, name, gender, age, city, signup_date"
1,products,100,4,4,"product_id, product_name, category, price"
2,orders,300,5,5,"order_id, customer_id, order_date, payment_met..."
3,order_items,764,5,5,"order_item_id, order_id, product_id, quantity,..."


In [51]:
for _, row in llm_dataset_summary.iterrows():
    print(f"- {row['dataset']}: {row['rows']}행 {row['columns']}열")
    print(f"  컬럼: {row['column_names']}")

- customers: 150행 6열
  컬럼: customer_id, name, gender, age, city, signup_date
- products: 100행 4열
  컬럼: product_id, product_name, category, price
- orders: 300행 5열
  컬럼: order_id, customer_id, order_date, payment_method, order_status
- order_items: 764행 5열
  컬럼: order_item_id, order_id, product_id, quantity, unit_price


- LLM에 제공한 Safe Context: 

온라인 쇼핑몰 데이터 분석을 시작하기 전에 다음 CSV 파일들의 구조를 이해하려고 합니다.

  데이터셋 요약:
  {llm_dataset_summary[['dataset', 'rows', 'columns', 'column_names']].to_string(index=False)}

  파일 간 관계:
  - customers.customer_id -> orders.customer_id
  - orders.order_id -> order_items.order_id
  - products.product_id -> order_items.product_id

  요청:
  1. 각 파일이 어떤 역할을 하는지 설명해 주세요.
  2. 분석 전에 확인해야 할 항목을 체크리스트로 정리해 주세요.
  3. 실제 데이터 확인 없이 단정한 내용과 추가 확인이 필요한 내용을 구분해 주세요.

  Read 1 file

- LLM이 제안한 추가 점검: 
  키의 유일성, 결측여부 / 참조 무결성 위반 여부
  dtype의 형식 및 이상치, 논리적 모순값 여부 등

- 실제 데이터에서 확인한 항목:
  "키의 실제 유일성·결측 여부, 참조 무결성 위반 여부"는 결측과 중복값 확인을 통하여 무결함을 확인

- 채택/수정/보류한 내용:
  "컬럼의 실제 dtype과 값 형식(가격이 문자열로 섞여 있을 수 있음)"을 보다 디테일하게 볼 필요가 있어 수정 채택

Evidence
![](images/05.png)

결과 관찰 
- claude code를 통하여 실습노트북에 있는 예시질문을 했고 각 파일의 역할과 분석 전 체크리스트, 단정한 내용과 추가확인 필요한 내용을 확인하였다.

나의 해석과 판단 
- 유용했던 부분 : "세 개의 키(customer_id, product_id, order_id)가 파일 간 연결 통로 역할을 한다" 이 부분은 앞선 항목에서 내가 주목했던 부분이고 LLM도 이 부분을 언급하고 있었다. 또한 분석 전 체크리스트를 상세하게 생성하여 주의할 점을 먼저 언급해주었다.
- 조심해야 하는 부분 : "컬럼의 실제 dtype과 값 형식(가격이 문자열로 섞여 있을 수 있음)"을 언급하였는데 문자로 인식되는 행이 따로 있었기 때문에 보다 주의깊게 데이터를 살펴 볼 필요가 있다.

업무·분석적 의미 
- LLM은 데이터를 직접 확인하지 않고 내가 생각했던 주요 포인트를 언급하였으며, 전처리 과정에서 필요한 체크리스트를 상세히 만들어주었다. 데이터 분석 전 상당히 유용한 자료로 활용 가능하다 판단된다.

한계와 추가 확인 사항 아직 
- 다만 데이터를 미리 본 상황이 아니기 때문에 디테일한 부분에서는 추가 정의,질문 등의 활용이 필요하다.

6. Chapter 03 최종 판단

데이터의 첫인상 3가지
1. 데이터의 전처리는 매우 중요하다.
2. 따라서 결측이나 중복 등의 오류가 있을 경우 어떻게 처리할지를 미리 정해야 한다.
3. 또한 price 숫자 컬럼도 콤마 등 문자가 섞일 수 있음을 예시로 확인되어 타입 전처리가 매우 중요함을 확인하였다.

다음 Chapter 전에 반드시 확인/처리해야 할 항목
1. LLM이 언급한 논리적 모순값(가입일 이전 주문) 등과 같은 결과는 확인해보지 않았다.
2. order_id + product_id 조합이 한 주문 안에서 중복되는지(같은 상품이 여러 줄로 쪼개져 있는지) 확인이 필요하다.

현재 데이터만으로 단정할 수 없는 것
 날짜 범위가 2025.09~2026.09 딱 1년치라서, 전년 대비 성장했는지 계절성이 있는지는 지금 데이터만으로는 알 수 없다.

[Chapter 03 Evidence]

1. CSV 파일 존재 여부
- customers.csv: PASS 
- products.csv: PASS 
- orders.csv: PASS 
- order_items.csv: PASS 

2. 데이터 규모
- customers: 150 rows × 6 cols
- products: 100 rows × 4 cols
- orders: 300 rows × 5 cols
- order_items: 764 rows × 5 cols

3. 품질 점검
- 주요 ID 결측: 0
- 주요 ID 중복: 0
- 전체 행 중복: 0
- 날짜 변환 실패: 0

4. 관계 검증
- 없는 customer_id: 0
- 없는 order_id: 0
- 없는 product_id: 0

5. LLM 활용
- 요청 목적: 실제 데이터를 열어보기 전, 요약 정보(행/열 수, 컬럼명, 파일 간 관계)만 주고 각 파일의 역할과 분석 전 체크리스트를 먼저 물어봄
- 실제 반영:   "컬럼의 실제 dtype과 값 형식(가격이 문자열로 섞여 있을 수 있음)"을 보다 디테일하게 볼 필요가 있어 수정 채택
- 사람이 검증한 항목: price 숫자칼럼에 오류부분을 변환

최종 제출 체크
 Notebook을 처음부터 끝까지 실행했습니다.
 오류 셀이 남아 있지 않습니다.
 핵심 Evidence를 첨부했습니다.
 관찰과 해석을 구분했습니다.
 개인정보/Secret이 없습니다.
 chapter03/chapter03.ipynb가 GitHub에서 정상 표시됩니다.
 최종 Notebook 파일 URL을 제출합니다.